### PR TITLE
EIP 1014 CREATE2

### LINKED ISSUE #356: [Constantinople] EIP 1014 CREATE2 Opcode
EIP: [EIP 1014](https://github.com/ethereum/EIPs/pull/1014) 
Work in progress: PR [#329](https://github.com/ethereumjs/ethereumjs-vm/pull/329)

Test Situation:
Many tests implemented
[Tests](https://github.com/ethereum/tests/tree/develop/GeneralStateTests/stCreate2) | [Filler](https://github.com/ethereum/tests/tree/develop/src/GeneralStateTestsFiller/stCreate2)

### PR DESCRIPTION
**Pull request by jwasinger**
Depends on https://github.com/ethereumjs/ethereumjs-util/pull/146.

Closes #356.

### CODE REVIEW

**Comment by holgerd77** [comment]
Hi Jared, I've now updated the ``ethereumjs-testing`` repo with the latest tests, this also includes your new ``CREATE2`` tests (see also https://github.com/ethereumjs/ethereumjs-vm/pull/348 for reference).

**Comment by holgerd77** [comment]
Tests are now updated to ``ethereumjs-testing`` ``v1.2.2`` (https://github.com/ethereumjs/ethereumjs-vm/pull/350), which includes the latest tests from the ``ethereum/tests`` repository with many new ``CREATE2`` test cases.

@jwasinger Are you planning to continue to work on this? Or is this also open for someone else to be continued (e.g. from Kyokan)?

**Comment by jwasinger** [comment]
Feel free to take this over.  I won't have time to complete it before devcon.

**Comment by rmeissner** [comment]
@holgerd77 @jwasinger as I am really interested in playing around with this feature, is there some ETA? Also is there some way I could contribute to get thisndone faster :smile:

**Comment by holgerd77** [comment]
Hi @rmeissner, as a first step I have now updated the according PR on the util library along the latest discussions from the EIP update PR [#1375](https://github.com/ethereum/EIPs/pull/1375) page, can you do a review of this?

I think one thing you can do is to follow the discussion alongside the PR linked above and come up with some examples for the hash creation - covering different cases - that would also help other implementations a lot.

You can further directly take on this PR if your time permits, that would be great! 😄 I have added you as collaborator to this library, so feel free to directly take on the PR here, rebase and re-submit.

**Comment by holgerd77** [comment]
Cool, pretty sure it won't be done before :smile:

**Comment by rmeissner** [comment]
The test cases don't included the correct gas costs yet. Also there are still quite some issues with the constantinople tests (the ci failures can be resolved if the new version if the utils package is released)

**Comment by holgerd77** [comment]
Just released ``ethereumjs-util`` [v6.0.0](https://github.com/ethereumjs/ethereumjs-util/releases/tag/v6.0.0) with the ``generateAddress2()`` helper function now included, sorry for the short delay.

**Comment by rmeissner** [comment]
It seems that just upgrading ethereumjs-utils from 5.2.0 to 6.0.0 is breaking quite some tests, if possible I would appreciate some help with that :)

**Comment by holgerd77** [comment]
Just looking into it.

File: lib/runCall.js [hunk 1]
```diff
@@ -5,6 +5,7 @@ const BN = ethUtil.BN
 const exceptions = require('./exceptions.js')
 
 const ERROR = exceptions.ERROR
+const EMPTY_CODE_HASH = ethUtil.keccak256()
```
Comment (holgerd77) [inline lib/runCall.js]: You are introducing an ethUtil call here, eventually that can also be a cause?

**Comment by holgerd77** [comment]
I've done a pure ``ethereumjs-util`` update PR https://github.com/ethereumjs/ethereumjs-vm/pull/369 to see if there are side/other effects or if it is really the library update.

**Comment by holgerd77** [comment]
Update: ah no, tests are also failing on the test PR, seems to be really the util update.

**Comment by holgerd77** [comment]
Ok, first update on https://github.com/ethereumjs/ethereumjs-vm/pull/369, if theory is correct the API tests should pass now.

**Comment by holgerd77** [comment]
API tests are passing, seems to me that I was on the correct track. However other tests are still failing. It just occurred to me that similar comparisons in the production code will likely now also fail.

The easier way to go would be eventually to update the merkle tree library to also use the ``v6.0.0`` util version.

Not sure about the way to go, directly fixing this would probably be the more correct way.

**Comment by holgerd77** [comment]
Ok, I found one left-over occurence of a sha3-named constant in StateManager, which have been removed in the v6.0.0 util version, this might very well be it, waiting for the tests to finish.

**Comment by holgerd77** [comment]
Ok, tests are working now, just merged the ``ethereumjs-util`` ``v6.0.0`` update PR to master! 😄 

Not sure, rebasing might be a bit tricky here, maybe do a local copy of your repo before.

**Comment by holgerd77** [comment]
Ok, great! :-) Now we have to wait for https://github.com/ethereumjs/ethereumjs-vm/pull/368 respectively an updated version - see current discussion on https://gitter.im/ethereum/tests since current blockchain tests are not behaving correctly. Then we can see how many additional tests from the Constantinople suite are passing.

**Comment by holgerd77** [comment]
(would you generally regard this as ready for testing?)

**Comment by rmeissner** [comment]
So right now I commented out some gas calculation, but in general it should be ready. I was running the create2 tests. Will check that again soon and give an update.

**Comment by holgerd77** [comment]
Ok, have merged the latest tests to master.

**Comment by holgerd77** [comment]
@rmeissner Maybe just ask your question directly on the tests Gitter channel at https://gitter.im/ethereum/tests, it is very well possible that the test cases are wrong itself, since test creation is very much on a late track this time.

Can you push a rebased version of your branch here? Don't dare to do it myself in the case that you have got local changes and I am eager to see the test results! 😄

**Comment by holgerd77** [comment]
Was this intended that you rebased on top of the SSTORE branch? Thought you would do it on master.

**Comment by rmeissner** [comment]
No it was not ... I wanted to test if this fixes any of the failing tests but the rebase back to master was not done properly ... sorry

**Comment by holgerd77** [comment]
Thanks 😄

**Comment by holgerd77** [comment]
Merged the ``SSTORE`` PR from @vpulim, if you want you can now "officially" rebase 😄, then overview on test failures should get a lot easier to grasp since this fixes a lot of tests.

**Comment by rmeissner** [comment]
hmmm the `test_gas_limit` are not failing locally

**Comment by holgerd77** [comment]
@rmeissner Are you sure you have the latest version of the ``ethereumjs-testing`` library installed? Anyhow, atm we can focus on the state tests since there are still current discussions on the tests repo on the blockchain tests consistency + updates in the works.

Would you generally regard this PR as open for review? Atm I have a tendency to merge rather quicker then later, then fixing still-failing tests on a general level gets easier. If there are some edge cases missing, we can also tackle these separately in subsequent PRs.

**Comment by holgerd77** [comment]
Ah ok, just had a closer look at the Constantinople state test results myself - sorry, could have done this before already - see that there are still some tests failing on the CREATE2 case.

**Comment by rmeissner** [comment]
@holgerd77 as far as I can see the test_gas_limit failures are not related to changes in this PR. For the 27 failing tests I think it comes down to gas usafe, so I need to figure that out (which is annoying since the success cases seem to work) :/

**Comment by holgerd77** [comment]
What tests are failing/passing if you do changes on this ``done(ERROR.INVALID_OPCODE)`` thing? This sounds like a general bug eventually?

**Comment by holgerd77** [comment]
For reference, currently the following tests are failing (``ethereumjs-testing`` ``v1.2.4``, ``95a3092``), extracted from [this](https://circleci.com/gh/ethereumjs/ethereumjs-vm/2447?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) CI run:

```shell
badOpcodes, 124
RevertInCreateInInitCreate2, 544
RevertOpcodeInCreateReturnsCreate2, 547
create2SmartInitCode, 559
create2collisionCode, 581-599
returndatacopy_afterFailing_create, 607
returndatacopy_following_revert_in_create, 610
RevertInCreateInInit, 2270
TouchToEmptyAccountRevert3, 2379
```

Summary:
```
1..4776
# tests 4776
# pass  4749
# fail  27
```

**Comment by holgerd77** [comment]
If someone else also wants to have a look, single test cases can be run with e.g.:

```shell
node ./tests/tester -s --fork='Constantinople' --test='RevertInCreateInInitCreate2'
```

Or e.g.:

```shell
node ./tests/tester -s --fork='Constantinople' --test='create2SmartInitCode' --data=1 --gas=0 --value=0
```

...for just selecting a single test case with the respective values for ``data``, ``gas`` and ``value``, see test [file](https://github.com/ethereum/tests/blob/develop/GeneralStateTests/stCreate2/create2SmartInitCode.json) and [filler](https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stCreate2/create2SmartInitCodeFiller.json) for reference over on the tests repository.

Values passed is just e.g. to ``data`` is just an index, e.g. if taken a list of data values like

```
"data" : [
	"0x0000000000000000000000000f572e5295c57f15886f9b263e2f6d2d6c7b5ec6",
	"0x0000000000000000000000001f572e5295c57f15886f9b263e2f6d2d6c7b5ec6"
]
```

then ``--data=1`` only means that the second value is taken. The filler also sometimes contains a comment what the test is about.

**Comment by holgerd77** [comment]
@rmeissner Did you manage to get tracing with aleth respectively testeth running? Just trying to do an own build of aleth (cause on the binary tracing is disabled), any known caveats? Or did you take a different path?

Will see how far I'll get. I also checked ``Py-EVM``, they don't have Constantinople state test support merged yet (see [this](https://github.com/ethereum/py-evm/pull/1181) PR), so no help from this side (yet).

**Comment by holgerd77** [comment]
@rmeissner Don't know if you've got more C++ insight: if I get "--vmtrace option requires a build with cmake -DVMTRACE=1" does this need to be applied to the config or build step? And is this one dash for the DVMTRACE option correct, or should this be rather ``--DVMTRACE=1``?

**Comment by holgerd77** [comment]
Oh, this was easy. Asked @cdetrio and he recommended to use ``go-ethereum``. This even worked without building from source, I had the ``evm`` tool used ready from my binary local geth install, test with trace can be run with:

```shell
evm --json --nomemory statetest ./GeneralStateTests/stCreate2/create2collisionCode2.json
```
Resulting in e.g. the following output (top: ethereumjs-vm, bottom: geth):

```shell
file: create2collisionCode2 test: create2collisionCode2
{"pc":0,"op":107,"gas":"0x545d4","gasCost":"0x3","stack":[],"depth":0,"opName":"PUSH12"}
{"pc":13,"op":96,"gas":"0x545d1","gasCost":"0x3","stack":["0x620102036000526003601df3"],"depth":0,"opName":"PUSH1"}
{"pc":15,"op":82,"gas":"0x545ce","gasCost":"0x3","stack":["0x620102036000526003601df3","0x0"],"depth":0,"opName":"MSTORE"}
{"pc":16,"op":96,"gas":"0x545c8","gasCost":"0x3","stack":[],"depth":0,"opName":"PUSH1"}
{"pc":18,"op":96,"gas":"0x545c5","gasCost":"0x3","stack":["0x0"],"depth":0,"opName":"PUSH1"}
{"pc":20,"op":96,"gas":"0x545c2","gasCost":"0x3","stack":["0x0","0xc"],"depth":0,"opName":"PUSH1"}
{"pc":22,"op":96,"gas":"0x545bf","gasCost":"0x3","stack":["0x0","0xc","0x14"],"depth":0,"opName":"PUSH1"}
{"pc":24,"op":245,"gas":"0x545bc","gasCost":"0x7d00","stack":["0x0","0xc","0x14","0x0"],"depth":0,"opName":"CREATE2"}
{"stateRoot":"82aa6fc100c9e661f594982b8ca46e67db59602049ab0cc2e31a55059ca4ae07"}

{"pc":0,"op":107,"gas":"0x545d4","gasCost":"0x3","memory":"0x","memSize":0,"stack":[],"depth":1,"opName":"PUSH12","error":""}
{"pc":13,"op":96,"gas":"0x545d1","gasCost":"0x3","memory":"0x","memSize":0,"stack":["0x620102036000526003601df3"],"depth":1,"opName":"PUSH1","error":""}
{"pc":15,"op":82,"gas":"0x545ce","gasCost":"0x6","memory":"0x","memSize":32,"stack":["0x620102036000526003601df3","0x0"],"depth":1,"opName":"MSTORE","error":""}
{"pc":16,"op":96,"gas":"0x545c8","gasCost":"0x3","memory":"0x","memSize":32,"stack":[],"depth":1,"opName":"PUSH1","error":""}
{"pc":18,"op":96,"gas":"0x545c5","gasCost":"0x3","memory":"0x","memSize":32,"stack":["0x0"],"depth":1,"opName":"PUSH1","error":""}
{"pc":20,"op":96,"gas":"0x545c2","gasCost":"0x3","memory":"0x","memSize":32,"stack":["0x0","0xc"],"depth":1,"opName":"PUSH1","error":""}
{"pc":22,"op":96,"gas":"0x545bf","gasCost":"0x3","memory":"0x","memSize":32,"stack":["0x0","0xc","0x14"],"depth":1,"opName":"PUSH1","error":""}
{"pc":24,"op":245,"gas":"0x545bc","gasCost":"0x7d00","memory":"0x","memSize":32,"stack":["0x0","0xc","0x14","0x0"],"depth":1,"opName":"CREATE2","error":""}
{"pc":25,"op":0,"gas":"0x1322","gasCost":"0x0","memory":"0x","memSize":32,"stack":["0x0"],"depth":1,"opName":"STOP","error":""}
{"output":"","gasUsed":"0x532b2","time":259029}
{"stateRoot": "29537ed68332a8b7e4c17f660f4c40b9a45c6deac425fdc80d170c5396900f48"}
```

**Comment by rmeissner** [comment]
@holgerd77 thanks for the update, will look into it tomorrow ... there is still some stuff to prepare for devcon 😉

**Comment by holgerd77** [comment]
The failures on collision are very likely not ``CREATE2`` specific (e.g. ``create2collisionCode`` test) but these cases are also not working under ``CREATE``, see the corresponding issue here: https://github.com/ethereumjs/ethereumjs-vm/issues/232

Will also try to dig a bit into that.

**Comment by holgerd77** [comment]
Rebased this to run with the latest tests.

**Comment by holgerd77** [comment]
Results with latest Constantinople state tests (``ethereumjs-testing`` ``v1.2.4``):

```shell
1..4776
# tests 4776
# pass  4749
# fail  27
```

**Comment by holgerd77** [comment]
Hi @rmeissner, tried to rebase this, this is currently conflicting with the latest ``StateManager`` refactoring changes from @mattdean-digicatapult introduced with this PR https://github.com/ethereumjs/ethereumjs-vm/pull/366/files#diff-466ee465f36ed5789ade4601e1b53f5b, so probably better if you have a look yourself and do a careful rebase once time permits.

**Comment by rmeissner** [comment]
@holgerd77 i will rebase it and fix it after devcon and try to figure out why all the create and create2 collision tests are failing

**Comment by rmeissner** [comment]
@holgerd77 I rebase and squashed the commits (to make future rebasing easier). I also fixed the linting tests. Will start looking into the create2/create tests

**Comment by rmeissner** [comment]
@holgerd77 I got it down to 4 failing tests. Tried to figure out what goes wrong but I think these are unrelated to create2.

E.g. `RevertInCreateInInitCreate2`

ethereumjs-vm
```
...
# {"pc":21,"op":96,"gas":"0x9fffea821","gasCost":"0x3","stack":["0x20"],"depth":0,"opName":"PUSH1"}
# {"pc":23,"op":85,"gas":"0x9fffea81e","gasCost":"0x0","stack":["0x20","0x0"],"depth":0,"opName":"SSTORE"}
# {"pc":24,"op":96,"gas":"0x9fffea756","gasCost":"0x3","stack":[],"depth":0,"opName":"PUSH1"}
...
```

vs `evm`
```
...
{"pc":21,"op":96,"gas":"0x9fffea821","gasCost":"0x3","memory":"0x","memSize":32,"stack":["0x20"],"depth":1,"refund":0,"opName":"PUSH1","error":""}
{"pc":23,"op":85,"gas":"0x9fffea81e","gasCost":"0x4e20","memory":"0x","memSize":32,"stack":["0x20","0x0"],"depth":1,"refund":0,"opName":"SSTORE","error":""}
{"pc":24,"op":96,"gas":"0x9fffe59fe","gasCost":"0x3","memory":"0x","memSize":32,"stack":[],"depth":1,"refund":0,"opName":"PUSH1","error":""}
...
```

The costs for `SSTORE` are different in both cases.

One side note the `gasCost` output of ethereumjs-vm is misleading since it doesn't include dynamic gas (e.g. hashing costs for create2 or the costs for sstore)

**Comment by rmeissner** [comment]
this might actually also fix https://github.com/ethereumjs/ethereumjs-vm/issues/232 didn't check everything but the example provided in the latest comment now passes

**Comment by holgerd77** [comment]
Rebased this.

**Comment by holgerd77** [comment]
If you are not actively intervening in the next half-an-hour or so, I would merge this 😄, would like to forward to @mattdean-digicatapult, hope this is ok.
